### PR TITLE
Theatre of Blood Session stats plugin

### PR DESCRIPTION
**Pull request by TheStonedTurtle**
Adds a plugin that tracks session stats for Theatre of Blood. This only tracks for the local player.

**NOTE: This data is lost when the client is closed**

### Tracked Stats

<details>
<p>

For each run, or **Attempt**, at the Theatre of Blood the following information will be tracked:

* Damage Dealt *(approximation)*
* Damage Taken
* Death Count
* Successfully Completed

Information is tracked by room, displaying the below information upon completion.
* Damage Dealt *(approximation)*
* Damage Taken

An Act is considered finished when you enter the next room, there is a total of 6 acts.

</p>
</details>

### Media

<details>
<p>

### In-Game messages:

</p>
</details>

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/theatreofblood/overlays/AttemptOverlay.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import net.runelite.client.plugins.theatreofblood.data.Attempt;
+import net.runelite.client.ui.overlay.Overlay;
+import net.runelite.client.ui.overlay.OverlayPosition;
+import net.runelite.client.ui.overlay.OverlayPriority;
+import net.runelite.client.ui.overlay.components.LayoutableRenderableEntity;
+import net.runelite.client.ui.overlay.components.LineComponent;
+import net.runelite.client.ui.overlay.components.PanelComponent;
+import net.runelite.client.ui.overlay.components.TitleComponent;
+
+public class AttemptOverlay extends Overlay
```
Comment (raiyni) [inline runelite-client/src/main/java/net/runelite/client/plugins/theatreofblood/overlays/AttemptOverlay.java]: These shouldn't be separate overlays

File: runelite-client/src/main/java/net/runelite/client/plugins/theatreofblood/data/Attempt.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+ */
+package net.runelite.client.plugins.theatreofblood.data;
+
+import java.util.ArrayList;
+import java.util.List;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+public class Attempt
```
Comment (raiyni) [inline runelite-client/src/main/java/net/runelite/client/plugins/theatreofblood/data/Attempt.java]: All of these data structures feel extraneous.

Comment (TheStonedTurtle) [reply runelite-client/src/main/java/net/runelite/client/plugins/theatreofblood/data/Attempt.java]: I'm trying to track by room and by attempt and this seemed like the easiest way, what would you recommend I do instead?

**Comment by raiyni** [comment]
Each attempt looks to be a sum of each room. Each room seems to historically current room. It also doesn't look like you are doing any advanced per room statistics. So I would guess a couplve v with enum keys would replace a lot of this boiler plate.

**Comment by MagicfTail** [comment]
Issues found from testing so far:

1. The raids completion % shows 1 instead of 100, idk if it's the same without 100% completion rate.
2. The death counter adds deaths to session doing 0 death raids, no deaths are added during the rooms themselves, so idk where they are coming from. 2 were added the first raid, and thereafter 1 every raid.

**Comment by MagicfTail** [comment]
3. The raid overlays show outside the rooms/theatre of blood area

**Comment by TheStonedTurtle** [comment]
> 
>     2. The death counter adds deaths to session doing 0 death raids, no deaths are added during the rooms themselves, so idk where they are coming from. 2 were added the first raid, and thereafter 1 every raid.

Deaths are currently being tracked by whenever the ToB Varbit changes to 3, it must change to 3 for some other reason as well, I'll have to change the way this is tracked.

**Comment by TheStonedTurtle** [comment]
@raiyni I've removed the overlays entirely and tried to simplify. Let me know if you see anything else.

@MagicfTail I believe I've fixed those issues, if you don't mind testing again lmk if you have any issues.

Thanks guys

**Comment by MagicfTail** [comment]
Completion % get's rounded to 0

**Comment by TheStonedTurtle** [comment]
> 
> 
> Completion % get's rounded to 0

I'm not sure how this is happening.. `(completions / attempts.size()) * 100` is what I'm doing to determine the percentage and I'm formatting it via `new DecimalFormat("#,###");'. Any ideas?

**Comment by TheStonedTurtle** [comment]
Removed the damage capping functionality as it messed with AOE attacks and wasn't that accurate anyway.

File: runelite-client/src/main/java/net/runelite/client/plugins/theatreofblood/TheatreOfBloodPlugin.java [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+	}
+
+	/**
+	 * Handles changes to the Theatre of Blood Varbit
+	 * @param old previous Theatre of Blood varbit value
+	 */
+	private void stateChanged(int old)
+	{
+		// TODO: Figure out a way to determine if they are logging back into a raid
+		// 0=default | 1=party | 2=inside/spectator | 3=dead-spectator
```
Comment (NathenSample) [inline runelite-client/src/main/java/net/runelite/client/plugins/theatreofblood/TheatreOfBloodPlugin.java]: Just to re-iterate the above - Once you're at the stage you're describing what your ints mean with comments, use an enum ;)

**Comment by TheStonedTurtle** [comment]
Working on making the plugin more generic for expand-ability, see #6452

**Pull request closed without merging**